### ISSUE TITLE
Use simple and clear React.js components

### ISSUE THREAD

**Issue by Ishirak** [versionpress/versionpress#1087]
This issue is part of a frontend code review, see #1083.

Most components in the front ende codebase don’t do one thing, but several at once. This is not a good technique to follow. Because of this any future changes or improvements are extremely hard and a chance to make a mistake is much higher.

For example, the [CommitsTable](https://github.com/versionpress/versionpress/blob/master/frontend/src/Commits/CommitsTable.react.tsx) component is a very large one. The component can be very hard to follow and its formating style can make the component unreadable. 
The component can be split into smaller ones to improve its readability and clarity. The component even uses [CommitsTableNote](https://github.com/versionpress/versionpress/blob/master/frontend/src/Commits/CommitsTableNote.react.tsx) component, which is much easier to follow and does just a one thing.

**Comment by vasek17** [comment]
Do you have any proposal about folder structure (or maybe naming convention) in this case?
If we would use small components `CommitsTableHeader`, `CommitsTableFooter`, `CommitsTableSelectAll` with current folder structure, the Commits folder would contain roughly 20-30 files which is not very readable.

I can think of two more solutions for this problem:
1. Small stateless components in the top of `CommitsTable.react.tsx` file - great for e.g. header or footer.
2. Multiple `render*` methods in the component - e.g. `renderHeader`, `renderFooter` - e.g. [here](https://github.com/versionpress/versionpress/blob/master/frontend/src/Commits/CommitsTable.react.tsx#L93)

@Ishirak: What do you think about them?

**Comment by vasek17** [comment]
I don't like titles like `Header` or `Footer` - I would prefer unique titles for components (at least for CSS classes). 
Also this could easily lead to very deep folder structures (4 levels are a lot imo). 
@borekb: What do you think about this?

**Comment by borekb** [comment]
- I'd prefer concrete names, not just `Header` and `Footer`, as well.
- I don't have a clear opinion when it comes to folder structure. Many files in a single folder feel bad but so do deeply nested folders. Maybe there could be some middle ground, e.g., having a `commitsTable` folder and then nested sub-components but always only one level deep. Is there some project that likely has the same problem and we could look at?
- I agree with the one component per file approach.

**Comment by vasek17** [comment]
> Is there some project that likely has the same problem and we could look at?

Quite common pattern is to use `components` folder with one or more related independent components. (e.g. [Calypso](https://github.com/Automattic/wp-calypso/tree/master/client/components)). It roughly corresponds to our folder [`common`](https://github.com/versionpress/versionpress/tree/master/frontend/src/common).

For other functionality (I don't know the exact rule on this) Calypso uses subfolders (e.g. `my-sites`, `me`, `post-editor`, ...) that contain max one level deep folders (excluding `docs` or `test` subfolders) with related components and corresponding controllers (e.g. [posts](https://github.com/Automattic/wp-calypso/tree/master/client/my-sites/posts), [menus](https://github.com/Automattic/wp-calypso/tree/master/client/my-sites/menus)). 

This looks like exactly the structure you described, I think it is pretty good for our scale.

**Comment by Ishirak** [comment]
Ok, I also don't like very deep folder structures, but sometimes it could be necessary. Nonetheless I don't think that it will be necessary in our case, so I'm proposing following structure which also considers #1091.

```JavaScript
frontend/src/
    config/        // Config files
    components/    // React.js components
        _common/
            Common1Component.tsx
            ...
        %component1NameFolder%/
            Component1Name.tsx
            maximumDeepLevelFolder/
                MaximumDeepLevelComponent.tsx
                ...
            ...
        %component2NameFolder%/
        ...
        %componentNNameFolder%/
    routes/        // React-router routes
    stores/        // MobX stores
    utils/         // API, ArrayUtils, ...
    ...
```

I hope that my proposal is clear enough, feel free to share your suggestions!

**Comment by borekb** [comment]
Looks good. Just two notes / questions:

1. Is camelCase a common convention? I think I've seen mostly dash-case (e.g., Calypso) but I haven't seen that many React projects I have to admit.
2. `_common` makes the folder appear at the top, I get that, but I'd still prefer just `common`. It makes for a more natural path to a component (`components/common/Component.tsx` rather than `components/_common/Component.tsx`).

BTW this also suggest dropping the `.react` part from `Component.react.tsx` which is something I like.

**Comment by Ishirak** [comment]
1. Every team has a different convention, Calypso has `dash-case`, our front end has `PascalCase`. Personally I preffer `PascalCase` just for React.js components files, everything else in `camelCase`. But it is just my opinion and it's definitely something I'm open about. If the rest of the team prefers something else, I won't have a problem with that decission at all.
2. Reasonable argument, I have no problem with this decission.
3. I think that dropping `.react` part from a React.js component name is more readable. All components are already identified by `.tsx` suffix, so there's no need to have `.react.tsx` suffix. Or am I missing something?

**Comment by borekb** [comment]
1. @vasek17, do you have a preference here?
2. Ok, let's go with `common` then.
3. No, it's a good suggestion, I like that. Let's drop the `.react` bit.

**Comment by vasek17** [comment]
1. I would go with dash-case, it is the most standard way in JS projects.
2. :+1:
3. Sure, it was added when TypeScript did not have a support for JSX syntax. No it doesn't add any value.

**Comment by Ishirak** [comment]
This issue is surely going to be a large one, so I'd like to discuss few details before the start of further implementation.

As an example of future use of simple and clear React.js component I chose the [BulkActionPanel](https://github.com/versionpress/versionpress/blob/master/frontend/src/BulkActionPanel/BulkActionPanel.react.tsx) component. For this component I've created a new [repository](https://github.com/Ishirak/new-vp-frontend). Its purpose is just to show my solution to this issue on the BulkActionPanel component, so that's why I didn't create a new branch. 

So that is a little backstory, now to the [solution](https://github.com/Ishirak/new-vp-frontend/tree/master/src/components/BulkActionPanel).

The component has been divided into a bunch of smaller, usually stateless components. The solution contains two possibilities of how to create a stateless component, but there exist a lot more possibilities, so if you prefer something different, feel free to share it.

- 1st:
 ```tsx
 interface ClearSelectionProps {
   changes: number;
    onClick(e: React.MouseEvent): void;
 }

 const ClearSelection: React.StatelessComponent<ClearSelectionProps> = ({ changes, onClick }) => ...

 export default ClearSelection;
 ```
- 2nd:
 ```tsx
 interface OptionProps {
   option: BulkActionPanelOption;
 }

 const Option = ( props: OptionProps ) => ...

 export default Option;
 ```

So a few questions, firstly about stateless components:

 1. **Which format do you prefer?**  Personally I like **2nd** – pure functions.

 2. **Should we just _default export_ components right away, like:**
 ```tsx
 interface OptionProps {
   option: BulkActionPanelOption;
 }

 export default ( props: OptionProps ) => ...
 ```

 3. **What is your opinion about small stateless components?** Personally I'm a huge fan of theirs. They just do one thing, are small, readable, testable and clear. One downside could be that in the codebase will be a lot of them, but a proper folder structure can solve this issue.

 Now a bunch of questions about the whole solution:

 4. **Is the solution clean, readable and maintainable?**

 5. **Do you have any issues with folder structure?**

 6. **Do you have any issues with `CamelCase` component names?** I prefer them, I've tried `dash-case`, but I don't like when _imports_ of components look like this:
 ```jsx
 import SomeComponent from 'some-component';
 ```

 7. **Is there something you do (not) like?**

 8. **Do you have any suggestions how to improve this solution?**

Btw, the BulkActionPanel solution counts on #1091 and so will be the entire solution to this issue.

**Comment by vasek17** [comment]
**Ad 5., 6.** 
I'd definitely use `PascalCase` for names of component files as the file name is the same as component title. But I am not sure about folder names (or "modules"/"namespaces").


1. 
  (**The current VersionPress way**) Folder and component in `PascalCase`. The folder has the same name as the "main" component (`import BulkActionPanel from '../BulkActionPanel/BulkActionPanel';`).
  ```JavaScript
    frontend/src/components/
      BulkActionPanel/
        BulkActionPanel.tsx
        Options/
          Options.tsx
        ...
      ...
    ...
  ```

2. 
  (**I would prefer this way**) Folder in `dash-case`, components in `PascalCase`. The folder does not have the same name as the "main" component (`import BulkActionPanel from '../bulk-action-panel/BulkActionPanel';`).
  ```JavaScript
    frontend/src/components/
      bulk-action-panel/
        BulkActionPanel.tsx
        options/
          Options.tsx
        ...
      ...
    ...
  ```

3. 
  (**The Calypso way (except the `PascalCase` component names)**) Folders in `dash-case`, the "main" component named `index.tsx`, other components in `PascalCase`. The main component has generic title, which is worse for "Go to file" in editor, however somehow better for imports in code (`import BulkActionPanel from '../bulk-action-panel';`).
  ```JavaScript
    frontend/src/components/
      bulk-action-panel/
        index.tsx
        options/
          index.tsx
          Option.tsx
        OtherComponent.tsx
        ...
      ...
    ...
  ```

4. 
  Folder and component in `PascalCase` except the "main" component which is named index. It has the most readable imports (`import BulkActionPanel from '../BulkActionPanel';`) but worse for "Go to file".
  ```JavaScript
    frontend/src/components/
      BulkActionPanel/
        index.tsx
        Options/
          index.tsx
          Option.tsx
        OtherComponent.tsx
        ...
      ...
    ...
  ```

**Comment by vasek17** [comment]
**Ad 1.** 
I'd go with the **first** solution (explicit type annotation `React.StatelessComponent<PropsType>`). It is mostly an aesthetic choise, the behaviour (even in IDEs) is mostly the same.

**Ad destructing props** - I think it is ok to destruct in function argument for up to 3 props (`({ changes, onClick })`). If there is more of them or the names are longer, I'd prefer (`(props)`) object in the function argument and destructing in function body.

**Ad 2.** 
IMO assigning to variable before export is more readable.

**Ad 3.** 
:+1:

**Comment by borekb** [comment]
@vasek17 Regarding 5&6, I'm quite strongly against `index.tsx` for the reasons you mentioned. Between 1 and 2, I don't have a strong preference. Dashes look slightly weird to me but maybe it's common in React projects, I don't know.

**Comment by vasek17** [comment]
**Ad 7., 8.**
**a)** I am not sure about the `interfaces` folder - do you think it is good to separate them from the "modules"? 
I would prefer one of those (preferably **2nd option**):

1. `Commits/Commits.d.ts`, `BulkActionPanel/BulkActionPanel.d.ts` - One definition file for one module 
2. `Commits/Commit.d.ts`, `BulkActionPanel/BulkActionPanelOption` - One definition file for one interface (or a few related)

**b)** I still prefer unique names for components - `BulkActionPanelTitle` instead of `Title`, ...

**c)** Just a little detail - I would move `routes.tsx` one folder up. I think there is no need for `routes` folder for only one file.

**d)** Where would the global or legacy **styles** be? I suggest `components/app/App.less` and `components/app/legacy/fileA-Z.less`.

**Comment by borekb** [comment]
The motivation for a) is probably that an interface might be useful in multiple modules. Is that the case, @Ishirak?

**Comment by vasek17** [comment]
@borekb 
> The motivation for a) is probably that an interface might be useful in multiple modules. Is that the case, @Ishirak?

The example repository uses both solutions (https://github.com/Ishirak/new-vp-frontend/blob/master/src/interfaces/StateInterfaces.d.ts and https://github.com/Ishirak/new-vp-frontend/blob/master/src/components/Commits/Commits.d.ts). 
I think the `Commit` interface would be also useful in multiple modules but somehow I think it should be in `commit(s)` module.

Maybe there will be some interface that would feel weird in any module, I can't think of any at the moment. If so, what about placing it to `common` directory?

**Comment by Ishirak** [comment]
First of all, thank you for the feedback, it's much appreciated.

## Directory structure

I don't have any issues with

```
frontend/src/components/
    bulk-action-panel/
      BulkActionPanel.tsx
      options/
        Options.tsx
      ...
    ...
  ...
```

and I'm with @borekb to **don't** use `index.tsx` as I share concerns as he does.

## Stateless components

I've only tried PhpStorm for the format of stateless components and therefore I've picked a solution which suits this IDE. But then after long time I gave a shot to the Visual Studio Code a was blown away how much it got better. Even the intellisense was **much** better for the TypeScript development. Therefore I don't have any issues with `React.StatelessComponent<PropsType>` format of writing Stateless components.

## Interface folder

@borekb 
> The motivation for a) is probably that an interface might be useful in multiple modules. Is that the case, @Ishirak?

That's exactly what I intended. I believe that if multiple components / state management library use a common interface, then it should be in a separate directory, but I'm new to TypeScript development, so in this case I really don't know what is the best solution. If the state management library uses an interface for managing state, I don't think that defining the interface in the component folder is a good idea.

## Component names

`BulkActionPanel`'s render method:

```jsx
render() {
    const { selectedCommits, enableActions } = this.props;
    const { options } = this.state;

    return (
      <div className='BulkActionPanel'>
        <div className='alignleft actions bulkactions'>
          <Title />
          <Options
            options={options}
            onChange={this.onSelectedValueChange}
          />
          <Submit
            onClick={this.onSubmitClick}
            isDisabled={!enableActions || selectedCommits.length === 0}
          />
          <ClearSelection
            changes={selectedCommits.length}
            onClick={this.onClearSelectionClick}
          />
        </div>
      </div>
    );
  }
```

vs

```jsx
render() {
    const { selectedCommits, enableActions } = this.props;
    const { options } = this.state;

    return (
      <div className='BulkActionPanel'>
        <div className='alignleft actions bulkactions'>
          <BulkActionPanelTitle />
          <BulkActionPanelOptions
            options={options}
            onChange={this.onSelectedValueChange}
          />
          <BulkActionPanelSubmit
            onClick={this.onSubmitClick}
            isDisabled={!enableActions || selectedCommits.length === 0}
          />
          <BulkActionPanelClearSelection
            changes={selectedCommits.length}
            onClick={this.onClearSelectionClick}
          />
        </div>
      </div>
    );
  }
```

I just don't believe that such a long name for a component is really necessary. I'm 100 % for such a name if a component is somehow common, but I don't belive that if a component is used just in the directory of its parent component that a such a long name is useful.

## React router – routes folder

In this case I'm thinking forward, what if we'd like to implement async file load, what if we'd like to add new feature which depends on a new route. I don't think that `routes` folder causes any harm, it's just an aesthetic choice. Components have a folder, config has a folder and even state management library will have a folder.

## Location of legacy styles
@vasek17 
> Where would the global or legacy styles be? I suggest `components/app/App.less` and `components/app/legacy/fileA-Z.less`.

:+1: 

## Location of Pages

In the core both pages are components, so I believe that their place is in the `components` directory, so `frontend/src/components/pages`?

**Comment by vasek17** [comment]
## Directory structure
Ok, let's agree on `dash-case` for folders and `PascalCase` for component files.

## Stateless components
:+1:

## Interface folder
I don't see the boundary between an interface that can be used in multiple modules and the other ones. I would choose either all interfaces in `interface` folder or all in modules.

## Component names
I agree. There is another way to do it: component name prefixed with module name and imported like this: `import Title from './BulkActionPanelTitle';`. I don't know if I like it.

I worry about the CSS classes that should be unique (and it is nice if they are the same as component's name). Maybe we can write them like this: `ModuleName-Component` and `ModuleName-Component-subelement` (e.g. `BulkActionPanel-Title`). 
I think I am quite ok with that.

OT: What is the current state of some better solution for modular styles in React components?

## routes folder
I think this is too much forward-looking but I agree it's only aesthetic choise. Let's keep the folder.

## Location of Pages
Yes, that's the third option. Pages are also components but the semantics if quite different. 
If we look at them as components, they shouldn't have their own `pages` folder but should be in corresponding module. If we look at them as pages, they shouldn't be in `components` folder. 

However, all three options are good and it is quite small detail. (I would go with the **first** option but it's up to you).

**Comment by Ishirak** [comment]
## Interface folder

I think that all interfaces in the `interface` folder isn't a bad idea. If there would be just interfaces which are used in the whole codebase, e.g. Commit and future MobX interfaces, that it could work well. Of course component's props definition interfaces would still remain in the component file.

## Component names
`import Title from './BulkActionPanelTitle';` is a solution, but I also don't like it.

Ad css classes:
If we would choose _small_ component names, such as `Title.tsx`, then I'm open for such a className.

Ad current style options:
[react-css-modules](https://github.com/gajus/react-css-modules) is very popular, and so is [aphrodite](https://github.com/Khan/aphrodite). I have never used them, but I'm open to them. CSS modules are gaining a lot of users recently, e.g. [twitter poll](https://twitter.com/mxstbr/status/756769255111192576). I think that both are viable solutions: they both use hash codes as a final CSS class attributes – no conflicts with WP or even another components.


## Location of Pages
You are right, in the `components` folder shouldn't be `pages` folder. But nonetheless they are component, so I think it's best to place them in the `src/components/home-page/`, `/not-found-page/` respectively.

**Comment by vasek17** [comment]
## Interface folder
OK, let's do it this way

## Component names
:+1: for *short* subcomponent names. In near future we can try to use CSS modules but it's not in the scope of this issue.

## Location of Pages
OK, but I would prefer only `src/component/home` and `src/component/not-found`.
Component names `HomePage` and `NotFoundPage`

**Comment by Ishirak** [comment]
**Awesome**, thank you for your insight and discussion.

**Comment by Ishirak** [comment]
[Here](https://github.com/versionpress/versionpress/tree/d5645b686ee0b47cd2de8d8ed5518fb3e63f9e83/frontend/src) is a preview of future directory structure. Only directory missing is `stores`, which will be added in the #1091. If you have any issues with this structure, please share your feedback.

**Comment by borekb** [comment]
:ship: Shipped in [4.0-alpha1](https://github.com/versionpress/versionpress/releases/tag/4.0-alpha1).
